pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -165,72 +166,79 @@ func (s *ServiceEntryStore) workloadEntryHandler(old, curr config.Config, event
-	// We will only select entries in the same namespace
-	entries := s.seWithSelectorByNamespace[curr.Namespace]
-	s.storeMutex.RUnlock()
-
-	// if there are no service entries, return now to avoid taking unnecessary locks
-	if len(entries) == 0 {
-		return
-	}
-	log.Debugf("Handle event %s for workload entry %s in namespace %s", event, curr.Name, curr.Namespace)
+	// includes instances new updated or unchanged, in other word it is the current state.
 	instancesUpdated := []*model.ServiceInstance{}
 	instancesDeleted := []*model.ServiceInstance{}
-	workloadLabels := labels.Collection{wle.Labels}
 	fullPush := false
 	configsUpdated := map[model.ConfigKey]struct{}{}
-	for _, se := range entries {
-		selected := false
-		if !workloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
-			if oldWle != nil {
-				oldWorkloadLabels := labels.Collection{oldWle.Labels}
-				if oldWorkloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
-					selected = true
-					instance := s.convertWorkloadEntryToServiceInstances(oldWle, se.services, se.entry, &key, s.Cluster())
-					instancesDeleted = append(instancesDeleted, instance...)
-				}
-			}
-		} else {
-			selected = true
-			instance := s.convertWorkloadEntryToServiceInstances(wle, se.services, se.entry, &key, s.Cluster())
-			instancesUpdated = append(instancesUpdated, instance...)
-		}
 
-		if selected {
-			// If serviceentry's resolution is DNS, make a full push
-			// TODO: maybe cds?
-			if se.entry.Resolution == networking.ServiceEntry_DNS || se.entry.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN {
-				fullPush = true
-				for key, value := range getUpdatedConfigs(se.services) {
-					configsUpdated[key] = value
-				}
+	addConfigs := func(se *networking.ServiceEntry, services []*model.Service) {
+		// If serviceentry's resolution is DNS, make a full push
+		// TODO: maybe cds?
+		if se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN {
+			fullPush = true
+			for key, value := range getUpdatedConfigs(services) {
+				configsUpdated[key] = value
 			}
 		}
 	}
 
-	if len(instancesDeleted) > 0 {
-		s.deleteExistingInstances(key, instancesDeleted)
+	cfgs, _ := s.store.List(gvk.ServiceEntry, curr.Namespace)
+	currSes := getWorkloadServiceEntries(cfgs, wle)
+	var oldSes map[types.NamespacedName]struct{}
+	if oldWle != nil {
+		if reflect.DeepEqual(oldWle.Labels, wle.Labels) {
+			oldSes = currSes
+		} else {
+			oldSes = getWorkloadServiceEntries(cfgs, oldWle)
+		}
 	}
-
-	if event != model.EventDelete {
-		s.updateExistingInstances(key, instancesUpdated)
+	selected, unSelected := compareServiceEntries(oldSes, currSes)
+	log.Debugf("workloadEntry %v select serviceEntry, selected %v, unSelected %v", selected, unSelected)
+	s.mutex.Lock()
+	for _, namespacedName := range selected {
+		services := s.services.getServices(namespacedName)
+		cfg := s.store.Get(gvk.ServiceEntry, namespacedName.Name, namespacedName.Namespace)
```

Comment:
I do not really understand this.

First, we list the SEs []config.Config. Then, we fetch all of the names/ns of the SEs that are relevant. Then, for each name/ns we look up the []config.Config... but we already had them before. Can't we just use the ones we already listed?

Reply:
I guess we are inside the lock now, but seems like we could have just `List` inside the lock?

At the very least this could use some comments explaining why we do it, if there is a good reason.

Reply:
Need an improvement, there is a race here, `Get` can return nil otherwise.